salmon/tagger/sources/base.py

```diff
@@ -95,31 +95,63 @@ def generate_track(
             **kwargs,
         }
 
+
     def determine_rls_type(self, data):
-        num_tracks = len(
-            list(chain.from_iterable([d.values() for d in data["tracks"].values()]))
-        )
-        if re.search(r"\bE\.?P\.?\b", data["title"], re.IGNORECASE):
-            return (
-                re.sub(r"\bE\.?P\.?\b", "", data["title"], flags=re.IGNORECASE),
-                "EP",
-            )
-        elif re.search(r"Single$", data["title"]):
-            return (
-                re.sub(r"-? *Single$", "", data["title"], flags=re.IGNORECASE),
-                "Single",
-            )
+        def strip_base_title(title):
+            return re.sub(r"\s*\(.*?\)", "", title).strip().lower()
+
+        tracks = [
+            track
+            for disc in data["tracks"].values()
+            for track in disc.values()
+        ]
+
+        num_tracks = len(tracks)
+        base_titles = {strip_base_title(track["title"]) for track in tracks}
+        main_artists = {artist[0] for artist in data.get("artists", []) if artist[1] == "main"}
+
+        title = data["title"]
+        rls_type = data.get("rls_type", "").lower()
```

Comment:
```suggestion
        rls_type = data.get("rls_type").lower() if data.get("rls_type") else ""
```

This needs a None check before calling `.lower()`. Was getting exception on sources not returning `rls_type`